### PR TITLE
[MLIR][TORCH] Add -convert-torch-to-tosa-custom pass for undecomposed selective ops

### PR DESCRIPTION
**Pull request by AmosLewis**
Use tosa::custom op to represent softmax op which we deliberately choose not to decompose into simple ops.The reason to do this is in this RFC here: https://github.com/llvm/torch-mlir/issues/1519

### CODE REVIEW

**Comment by ramiro050** [comment]
The CI error is due to this patch. The program is crashing when creating the `tosa::CustomOp`:

https://github.com/llvm/torch-mlir/actions/runs/3292098619/jobs/5427059644#step:6:6218

**Comment by ramiro050** [comment]
I don't think this is the right approach for what this patch is trying to achieve. As a user of torch-mlir, I would expect that running the `torch-to-tosa` pass on any set of supported ops will result in standard tosa ops. Users should not be expected by default to handle arbitrary custom ops. To use custom ops should be a choice made by the user.

I think this requires a more general solution where users can specify which torch-ops to pass down as custom ops, and have the conversion to `tosa::CustomOp` happen automatically. Since this will require some non-trivial changes to several parts of torch-mlir, can you open an RFC issue with what you're tryinig to achieve, and we can brainstorm on the best solution.

**Comment by AmosLewis** [comment]
> The CI error is due to this patch. The program is crashing when creating the `tosa::CustomOp`:
> 
> https://github.com/llvm/torch-mlir/actions/runs/3292098619/jobs/5427059644#step:6:6218

The interesting thing is when I run [https://github.com/llvm/torch-mlir/actions/runs/3292098619/jobs/5427059644#step:6:6218:~:text=6191-,%2D%2D,%2D%2D,-6194](https://github.com/llvm/torch-mlir/actions/runs/3292098619/jobs/5427059644#step:6:6218:~:text=6191-,%2D%2D,%2D%2D,-6194) on my local machine, everything looks good. Here is my local run cmd print:
```
➜  torch-mlir git:(tosa-custom-softmax) ✗ torch-mlir-opt test/Conversion/TorchToTosa/basic.mlir -convert-torch-to-tosa -split-input-file -verify-diagnostics | FileCheck  test/Conversion/TorchToTosa/basic.mlir
➜  torch-mlir git:(tosa-custom-softmax) ✗ 
```

**Comment by AmosLewis** [comment]
> 

Yes. The RFC is on the way by my partner. I just try to do something before the RFC is done. This is the initial step for demonstrating custom op could be used to handle complex ops that we don't want to decompose.  There should be a pass/list where we add ops and only ops from this list should go through the custom op route, the rest of op should be lower normally.

**Comment by silvasean** [comment]
Let's wait for the RFC to discuss more. Per-op custom patterns are definitely not the right way to do this. Maybe we could have a pass `torch-to-tosa-convert-to-custom{op-list=aten.softmax.int,...}` which systematically converts any Torch op to the appropriate tosa.custom.

**Comment by sjarus** [comment]
This is an interesting PR. Integer domain softmax _could_ be done using a custom op, but need not be. There is a working and bit accurate decomposition of TFLite softmax that is fairly detailed but works: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc#L1375

Please note that the legalize_common.cc there is the same as TosaLegalizeCommon.cc - the goal here is to eventually move these builders into the llvm-project TOSA codebase. 

The straightforward solution here is to copy the convertSoftMaxOp() into TosaLegalizeCommon.cpp, then just call that. It should do all the work for you. This approach is used by several other TorchToTosa legalizations that invoke calls to TosaLegalizeCommon to implement legalizations using previously solved pattern replacements.

**Review by sjarus — CHANGES_REQUESTED** [review]
Please see previous comment.

**Comment by AmosLewis** [comment]
> Let's wait for the RFC to discuss more. Per-op custom patterns are definitely not the right way to do this. Maybe we could have a pass `torch-to-tosa-convert-to-custom{op-list=aten.softmax.int,...}` which systematically converts any Torch op to the appropriate tosa.custom.

The RFC is ready here. [https://github.com/llvm/torch-mlir/issues/1519](https://github.com/llvm/torch-mlir/issues/1519) @ramiro050 @sjarus

**Comment by sjarus** [comment]
Adding @eric-k256

**Comment by AmosLewis** [comment]
Here is my most recent local result:
```
➜  torch-mlir git:(tosa-custom-softmax) ✗ torch-mlir-opt -convert-torch-to-tosa="custom-ops=torch.aten.softmax.int"    test/Conversion/TorchToTosa/custom.mlir  | externals/llvm-project/mlir/utils/generate-test-checks.py

// NOTE: Assertions have been autogenerated by utils/generate-test-checks.py

// The script is designed to make adding checks to
// a test case fast, it is *not* designed to be authoritative
// about what constitutes a good test! The CHECK should be
// minimized and named to reflect the test intent.



// CHECK-LABEL:   func.func @torch.aten.softmax.int$cst_dim(
// CHECK-SAME:                                              %[[VAL_0:.*]]: !torch.vtensor<[2,3],f32>) -> !torch.vtensor<[2,3],f32> {
// CHECK:           %[[VAL_1:.*]] = torch_c.to_builtin_tensor %[[VAL_0]] : !torch.vtensor<[2,3],f32> -> tensor<2x3xf32>
// CHECK:           %[[VAL_2:.*]] = torch.constant.none
// CHECK:           %[[VAL_3:.*]] = torch.constant.int 1
// CHECK:           %[[VAL_4:.*]] = "tosa.custom"(%[[VAL_1]]) {dim = 1 : i64, identifier = "softmax"} : (tensor<2x3xf32>) -> tensor<2x3xf32>
// CHECK:           %[[VAL_5:.*]] = torch_c.from_builtin_tensor %[[VAL_4]] : tensor<2x3xf32> -> !torch.vtensor<[2,3],f32>
// CHECK:           return %[[VAL_5]] : !torch.vtensor<[2,3],f32>
// CHECK:         }
```

**Comment by AmosLewis** [comment]
Just got a E2E test bug about the option I added:
```
python -m e2e_testing.main --config=tosa -v
: CommandLine Error: Option 'custom-ops' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
```

File: include/torch-mlir/Dialect/TorchConversion/Transforms/Passes.h
Comment (ramiro050) [inline include/torch-mlir/Dialect/TorchConversion/Transforms/Passes.h:41]: I think the build issue is here. There are two options both with the same name `"custom-ops"`

Comment (AmosLewis) [reply include/torch-mlir/Dialect/TorchConversion/Transforms/Passes.h]: Make sense. Thanks, Ramiro. I have this fixed, could you review and approved this patch if everything else makes sense?

**Comment by AmosLewis** [comment]
> Please see previous comment.
Please review this patch again. The purpose of this patch is to work with the op that we choose to not decompose.

**Comment by AmosLewis** [comment]
> Please see previous comment.

Please review this patch again. The purpose of this patch is to work with the op that we choose to not decompose.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp
Comment (eric-k256) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:3905]: This feels awkward, and isn't checking the value of the customOps to make sure that softmax is in the customOp list. I think a better option is to always insert the pattern, and then at the top of the softmax matchAndRewrite, search for torch.aten.softmax.int in the string, and use that to determine whether to create the custom op. 

Some checking that the customOp list is valid would be good. With this change, someone could put any op in the customOp list and get no warning that the pass will not do what is being asked. At least having a list of ops that support legalization to tosa.custom could be used to warn if an op was passed that wasn't expected.

File: include/torch-mlir/Dialect/TorchConversion/Transforms/Passes.h
Comment (eric-k256) [inline include/torch-mlir/Dialect/TorchConversion/Transforms/Passes.h:35]: I don't see any other reference to this, is this option needed, or is the custom-ops ListOption sufficient?

**Review by eric-k256 — COMMENTED** [review]
This feels halfway between being a one-off solution for torch.aten.softmax.int, and providing a flexible way to handle many ops in this manner. With some relatively small changes, I think it can be an example use of the flexible mechanism.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:3032]: I think there is still more work to be done before I would be comfortable merging this PR. The first reason is that this creates a precedent of allowing per-op custom patterns for converting to `tosa::CustomOp`. Other contributors of torch-mlir who might need a different op lowered to `tosa::CustomOp` might follow this approach, and we definitely don't want future PRs adding more per-op custom patterns to `tosa::CustomOp`.

The second reason is that this is not yet a solution that can be generalized nicely to other ops. Two main reasons come to mind at the moment:

1.  How would this approach work if there was also a conversion pattern for `AtenSoftmaxIntOp` that converts to standard tosa ops? Which pattern would get applied when both patterns are active? This is where having a separate pass for custom ops simplifies things.

2. This does not address one of the issues raised in the RFC, which is the handling of types such as `!torch.none`. This PR simply ignores the third argument of `AtenSoftmaxIntOp`. If users start depending on this convention for this op, it will make it harder to change things in the future. We need to first come up with a good design for the rules of conversion for all types that users can then depend on. Essentially, we need to clearly define a `tosa::CustomOp` contract. For example, are we always going to convert constant `int`s to rank-1 tensors as is done in this PR, or should we use rank-0 tensors or encode the int as an attribute? This is crucial to get right in the beginning so that users can have something stable to depend on.

I think we should continue iterating on this.

**Review by ramiro050 — CHANGES_REQUESTED** [review]

**Comment by sjarus** [comment]
> This feels halfway between being a one-off solution for torch.aten.softmax.int, and providing a flexible way to handle many ops in this manner. With some relatively small changes, I think it can be an example use of the flexible mechanism.

Also, isn't custom() being deprecated or at least being replaced with interface() aligned with ml_program ?

I think that torch softmax quantized is better off leveraging the existing functional lowering into TosaLegalizeCommon that is bit accurate for tfl.softmax.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3032]: Can we split it into small pieces with several PR? We are not trying to implement everything of https://github.com/llvm/torch-mlir/issues/1519 in one PR.

Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3032]: > I think there is still more work to be done before I would be comfortable merging this PR. The first reason is that this creates a precedent of allowing per-op custom patterns for converting to `tosa::CustomOp`. Other contributors of torch-mlir who might need a different op lowered to `tosa::CustomOp` might follow this approach, and we definitely don't want future PRs adding more per-op custom patterns to `tosa::CustomOp`.
> 
> The second reason is that this is not yet a solution that can be generalized nicely to other ops. Two main reasons come to mind at the moment:
> 
> 1. How would this approach work if there was also a conversion pattern for `AtenSoftmaxIntOp` that converts to standard tosa ops? Which pattern would get applied when both patterns are active? This is where having a separate pass for custom ops simplifies things.
> 2. This does not address one of the issues raised in the RFC, which is the handling of types such as `!torch.none`. This PR simply ignores the third argument of `AtenSoftmaxIntOp`. If users start depending on this convention for this op, it will make it harder to change things in the future. We need to first come up with a good design for the rules of conversion for all types that users can then depend on. Essentially, we need to clearly define a `tosa::CustomOp` contract. For example, are we always going to convert constant `int`s to rank-1 tensors as is done in this PR, or should we use rank-0 tensors or encode the int as an attribute? This is crucial to get right in the beginning so that users can have something stable to depend on.
> 
> I think we should continue iterating on this.

Can we split it into small pieces with several PR? We are not trying to implement everything of https://github.com/llvm/torch-mlir/issues/1519 in one PR.

Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3032]: Can we split it into small pieces with several PR? We are not trying to implement everything of https://github.com/llvm/torch-mlir/issues/1519 in one PR.

Comment (ramiro050) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3032]: Sure! It's fine to implement in chunks with unimplemented paths resulting in an error with a good error message. Given that the first PR will lay down the structure that future PRs will build on, for this RFC the first PR should include a document describing the `tosa::CustomOp` contract as well as create the `TorchToTosaCustom` pass, and have a simple working example of an op being systematically converted to `tosa::CustomOp`.

File: include/torch-mlir/Dialect/TorchConversion/Transforms/Passes.h
Comment (AmosLewis) [reply include/torch-mlir/Dialect/TorchConversion/Transforms/Passes.h]: Deleted

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Cool, I just create a new ConvertTorchToTosaCustom pass and add the document in the description. The example is also there.

File: include/torch-mlir/Dialect/TorchConversion/Transforms/Passes.h
Comment (ramiro050) [inline include/torch-mlir/Dialect/TorchConversion/Transforms/Passes.h:35]: nit: I think it's a bit clearer to say "List of ops to be converted to tosa::CustomOp"

File: lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp:45]: This is not a systematic conversion to `tosa::CustomOp`. By systematic, I mean that the pattern should apply to every op in the IR and it should convert to `tosa::CustomOp` without using any op-specific knowledge. In other words, the transformations should be done on `Operation *` objects.

File: lib/Dialect/TorchConversion/Transforms/Passes.cpp
Comment (ramiro050) [inline lib/Dialect/TorchConversion/Transforms/Passes.cpp:100]: nit: run `git clang-format`. Spacing is off

File: include/torch-mlir/Conversion/Passes.td
Comment (ramiro050) [inline include/torch-mlir/Conversion/Passes.td:120]: This is not a contract. A contract is something that describes in detail exactly what the expectations are before or after some transformation happens. For example, the backend contract in torch-mlir describes what the expectations are after conversion to torch backend is finished: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#the-backend-contract.

In this case, the contract should outline what each of the different torch dialect types will get turned into when this pass is finished. It should also outline exactly in what way attribute strings will be used (if they are used at all) to encode arguments. This document is meant to help users understand exactly what will happen when this pass takes place from just looking at the aten op signatures. Otherwise, there is no way for users to reliably know how to further lower `tosa::CustomOp` to their backend.

**Review by ramiro050 — DISMISSED** [review]

**Comment by ramiro050** [comment]
> > This feels halfway between being a one-off solution for torch.aten.softmax.int, and providing a flexible way to handle many ops in this manner. With some relatively small changes, I think it can be an example use of the flexible mechanism.
> 
> Also, isn't custom() being deprecated or at least being replaced with interface() aligned with ml_program ?
> 
> I think that torch softmax quantized is better off leveraging the existing functional lowering into TosaLegalizeCommon that is bit accurate for tfl.softmax.

@AmosLewis, can you address this?

File: lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp:46]: I thought we can do the #define INSERT_ATENOP_PATTERN(AtenOp) ...     kind of thing for every op just like what the TorchToTosa.cpp has done.

File: include/torch-mlir/Dialect/TorchConversion/Transforms/Passes.h
Comment (AmosLewis) [reply include/torch-mlir/Dialect/TorchConversion/Transforms/Passes.h]: Done

File: lib/Dialect/TorchConversion/Transforms/Passes.cpp
Comment (AmosLewis) [reply lib/Dialect/TorchConversion/Transforms/Passes.cpp:100]:
```
(mlir_venv) nod% git clang-format .
clang-format did not modify any files
```

Comment (ramiro050) [reply lib/Dialect/TorchConversion/Transforms/Passes.cpp:100]: `git clang-format` will only modify staged files. To modify the files already committed, you need to specify the range of commits. For example: `git clang-format HEAD~1`

File: lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp
Comment (ramiro050) [reply lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp:46]: No, we don't want to write a new pattern for every possible PyTorch op (this is what I meant in my previous comment when I mentioned that we don't want "per-op custom patterns for converting to tosa::CustomOp"). This type of transformation should be achievable with a single pattern that applies to all ops.

Comment (ramiro050) [reply lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp:46]: > And Since the !torch.none kind of type conversion is not handled yet, and 'tosa.custom' op operand must be a tensor of number values.

If this PR does not yet plan on handling `!torch.none`, then you should choose a simpler op that does not take `!torch.none`s as inputs for the first PR.

> I don't think there is a general way to handle all ops.

If there is no general way of converting a good subset of PyTorch ops to `tosa::CustomOp`, then we should improve the interface to `tosa::CustomOp` first, otherwise this will bring little value to users. However, I do think you should already be able to handle a good subset of ops, namely, all ops that at compile time take either tensors as input or constants. I would focus on this and reject all other ops for now.

**Comment by eric-k256** [comment]
> 

We had thoughts of making bigger changes to tosa.custom, and calling it tosa.interface. That hasn't happened, but I do have a proposed change to tosa.custom in the dialect here: https://reviews.llvm.org/D137133
I think that makes the generic Operation* mapping simpler, as you can now encode attributes into the implementation_attrs 

The other point Suraj was making is that we do have a legalization for softmax int into a series of TOSA ops that could be used to work aorund this specific problem. The larger problem of conversion from torch to tosa.custom remains, which the pass described by Ramiro aims to solve.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Done

File: lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp:46]: Any suggestion how to handle the **!torch.none**?

File: lib/Dialect/TorchConversion/Transforms/Passes.cpp
Comment (AmosLewis) [reply lib/Dialect/TorchConversion/Transforms/Passes.cpp]: Done

File: include/torch-mlir/Conversion/Passes.td
Comment (AmosLewis) [reply include/torch-mlir/Conversion/Passes.td:120]: Done

**Comment by AmosLewis** [comment]
@ramiro050  I just added a generic MatchAndRewrite for torch-to-tosa-custom pass. No need to add write new MatchAndRewrite for new ops. It worked for the 2 examples in custom.mlir. Please review. We can add more type conversion support to it later. Please review.

File: include/torch-mlir/Conversion/Passes.td
Comment (ramiro050) [inline include/torch-mlir/Conversion/Passes.td:120]: nit: `donnot` -> `do not`

File: include/torch-mlir/Conversion/Passes.td
Comment (ramiro050) [inline include/torch-mlir/Conversion/Passes.td:122]: `will used` -> `will be used`

File: include/torch-mlir/Conversion/Passes.td
Comment (ramiro050) [inline include/torch-mlir/Conversion/Passes.td:123]: `where is the` -> `where the`

File: include/torch-mlir/Conversion/Passes.td
Comment (ramiro050) [inline include/torch-mlir/Conversion/Passes.td:128]: nit: I would replace `?` with `AnySize`, since `?` means the type has a single dynamic dimension

File: include/torch-mlir/Conversion/Passes.td
Comment (ramiro050) [inline include/torch-mlir/Conversion/Passes.td:126]: I don't think this last sentence is necessary. It's more of an implementation detail.

File: include/torch-mlir/Conversion/Passes.td
Comment (ramiro050) [inline include/torch-mlir/Conversion/Passes.td:131]: I would also add a TODO explaining the things that are currently unsupported.

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: include/torch-mlir/Conversion/Passes.td
Comment (AmosLewis) [reply include/torch-mlir/Conversion/Passes.td:122]: Done

File: include/torch-mlir/Conversion/Passes.td
Comment (AmosLewis) [reply include/torch-mlir/Conversion/Passes.td:120]: Done

File: lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp:74]: Done

File: lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp:63]: Done

File: lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp:45]: Done

File: include/torch-mlir/Conversion/Passes.td
Comment (AmosLewis) [reply include/torch-mlir/Conversion/Passes.td:123]: Done

File: include/torch-mlir/Conversion/Passes.td
Comment (AmosLewis) [reply include/torch-mlir/Conversion/Passes.td:128]: Done

File: include/torch-mlir/Conversion/Passes.td
Comment (AmosLewis) [reply include/torch-mlir/Conversion/Passes.td:126]: Done

File: include/torch-mlir/Conversion/Passes.td
Comment (AmosLewis) [reply include/torch-mlir/Conversion/Passes.td:131]: Done

**Comment by Svoch** [comment]
Folks, I believe the approach I have in my recent PR (which is in reference to the discussion in [this Selective Op decomposition RFC](https://github.com/llvm/torch-mlir/issues/1519)) addresses the outstanding comments in this PR.

@AmosLewis - do you mind if we move the discussion over to [the other PR](https://github.com/llvm/torch-mlir/pull/1563)?

**Comment by AmosLewis** [comment]
> Folks, I believe the approach I have in my recent PR (which is in reference to the discussion in [this Selective Op decomposition RFC](https://github.com/llvm/torch-mlir/issues/1519)) addresses the outstanding comments in this PR.
> 
> @AmosLewis - do you mind if we move the discussion over to [the other PR](https://github.com/llvm/torch-mlir/pull/1563)?
Sure, Let's relocate to your patch.

**Pull request closed without merging**
